File: microsoft/KClosestPoints.java
import java.io.*;
import java.util.*;
import java.math.BigInteger;

/*
 * Write a function that takes three parameters, 
 * as arguments, Points[], vertex and k
 * and returns k closest Points.
 *
 * Example:
 * points[] = [{1,2}, {3,-1}, {1,4}]
 * vertex = {2,2}
 * k = 2
 * Return k closest points from points[]
 * 
 * Naive approach: T:O(N*lgN) S:O(N)
 *  - For each point in points:
 *    - calculate distance b/w point and vertex.
 *    - store in result array
 *  - Sort the result array w.r.t distance.
 *  - Return first k elements.
 *
 * Better approach: T:O(N*lgK) S:O(K)
 *  - Create a MaxHeap.
 *  - For each point in points:
 *    - calculate distance b/w point and vertex.
 *    - If heap size is less than k:
 *      - push to heap. Heapify.
 *    - If heap size is k:
 *      - If dist < root, heapify. Else continue.
 *  - Return heap;
 */

class Solution {
  public static class Point{
    int x;
    int y;
    Point(int x, int y) {
      this.x = x;
      this.y = y;
    }
    void print() {
      System.out.println("Point {" + x + ", " + y + "}");
    }
  }
  
  public static class Distance implements Comparable<Distance>{
    Integer dist;
    Integer idx;
    Distance(Integer dist, Integer idx){
      this.dist = dist;
      this.idx = idx;
    }
    
    @Override
    public int compareTo(Distance obj) {
      return this.dist.compareTo(obj.dist);
    }
  }
  
  public static int calculateDist(Point point, Point vertex) {
    int x_diff = Math.abs(point.x - vertex.x);
    int y_diff = Math.abs(point.y - vertex.y);
    return (x_diff * x_diff) + (y_diff * y_diff);
  }
  
  public static Point[] kClosestPoints1(Point[] points, Point vertex, int k) {
    if (k<=0) {
      return new Point[0];
    }
    if (k >= points.length) {
      return points;
    }
    Distance[] sqDistList = new Distance[points.length];
    for(int i=0; i<points.length; i++) {
      int dist = calculateDist(points[i], vertex);
      sqDistList[i] = new Distance(dist, i);
    }
    
    Arrays.sort(sqDistList);
    for(int i=0; i<sqDistList.length; i++) {
      System.out.println("id: " + sqDistList[i].idx + " dist: " + sqDistList[i].dist);
    }
    
    Distance[] result = Arrays.copyOfRange(sqDistList, 0, k);
    Point[] kClosestPoints = new Point[result.length];
    for(int i=0; i<result.length; i++){
      kClosestPoints[i] = points[result[i].idx];
    }
    return kClosestPoints;
  }
  
  public static void main(String[] args) {
    Point p1 = new Point(1,2);
    Point p2 = new Point(4,-3);
    Point p3 = new Point(1,4);
    Point[] points = new Point[]{p1, p2, p3};
    Point vertex = new Point(2, 2);
    
    Point[] result = kClosestPoints1(points, vertex, 4);
    for(int i=0; i<result.length; i++) {
      result[i].print();
    }
  }
}

File: microsoft/KClosestPoints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

/*
 * Write a function that takes three parameters, 
 * as arguments, Points[], vertex and k
 * and returns k closest Points.
 *
 * Example:
 * points[] = [{1,2}, {3,-1}, {1,4}]
 * vertex = {2,2}
 * k = 2
 * Return k closest points from points[]
 * 
 * Naive approach: T:O(N*lgN) S:O(N)
 *  - For each point in points:
 *    - calculate distance b/w point and vertex.
 *    - store in result array
 *  - Sort the result array w.r.t distance.
 *  - Return first k elements.
 *
 * Better approach: T:O(N*lgK) S:O(K)
 *  - Create a MaxHeap.
 *  - For each point in points:
 *    - calculate distance b/w point and vertex.
 *    - If heap size is less than k:
 *      - push to heap. Heapify.
 *    - If heap size is k:
 *      - If dist < root, heapify. Else continue.
 *  - Return heap;
 */

typedef struct point_t{
  int x;
  int y;
}Point;

typedef struct distance_t{
  uint dist;
  uint idx;
}Distance;

void swap(Distance& d1, Distance& d2) {
  Distance temp = d1;
  d1 = d2;
  d2 = temp;
}

void maxHeapify(vector<Distance>& distList, uint idx) {
  uint largest = idx;
  uint left = 2*idx + 1;
  uint right = 2*idx + 2;
  if (left < distList.size() && distList[left].dist > distList[largest].dist) {
    largest = left;
  }
  if (right < distList.size() && distList[right].dist > distList[largest].dist) {
    largest = right;
  }
  
  if (largest != idx) {
    swap(distList[idx], distList[largest]);
    maxHeapify(distList, largest);
  }
}

bool distCompare(Distance const& d1, Distance const& d2) {
  if(d1.dist == d2.dist) {
    return false;
  }
  return d1.dist > d2.dist ? false:true;
}

void printDist(vector<Distance> const& distList){
  for(uint i=0; i<distList.size(); i++) {
    cout << "Distance: {" << distList[i].dist << ", " << distList[i].idx << "}" << endl;
  }
}

void printPoints(vector<Point> const& points) {
  for(uint i=0; i<points.size(); i++) {
    cout << "Point: {" << points[i].x << ", " << points[i].y << "}" << endl;
  }
}

uint calcDist(Point p, Point v) {
  uint x_diff = abs(p.x - v.x);
  uint y_diff = abs(p.y - v.y);
  return (x_diff * x_diff) + (y_diff * y_diff);
}

vector<Point> kClosestPoints1(vector<Point> points, Point vertex, uint k) {
  vector<Point> result;
  vector<Distance> distList;
  for(uint i=0; i<points.size(); i++) {
    uint dist = calcDist(points[i], vertex);
    distList.push_back({dist, i});
  }
  sort(distList.begin(), distList.end(), &distCompare);
  printDist(distList);
  for(uint i=0; i<k && k<=distList.size(); i++) {
    result.push_back(points[distList[i].idx]);
  }
  return result;
}

vector<Point> kClosestPoints2(vector<Point> points, Point vertex, uint k) {
  vector<Point> result;
  vector<Distance> distList;
  for(uint i=0; i<points.size(); i++) {
    uint dist = calcDist(points[i], vertex);
    if (distList.size() < k) {
      distList.push_back({dist, i});
      maxHeapify(distList, 0);
    } else {
      if (distList[0].dist > dist) {
        distList[0] = {dist, i};
        maxHeapify(distList, 0);
      }
    }
  }
  for(uint i=0; i < distList.size(); i++) {
    result.push_back(points[distList[i].idx]);
  }
  return result;
}

// To execute C++, please define "int main()"
int main() {
  Point p1 = {1, 2};
  Point p2 = {4, -12};
  Point p3 = {1, 6};
  Point p4 = {10, 15};
  vector<Point> points;
  points.push_back(p1);
  points.push_back(p2);
  points.push_back(p3);
  points.push_back(p4);
  
  vector<Point> result = kClosestPoints2(points, {2,2}, 3);
  printPoints(result);
  return 0;
}
